### PR TITLE
fix(postgres-cdc): only prepare statement in the backfill process

### PR DESCRIPTION
**Pull request by StrikeW**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Move the prepare statement step to snapshot read, so that we don't need to prepare the select statement if the table has finished backfill. And can avoid problem caused by upstream #17807
- currently mysql cdc doesn't need to prepare statement, since we don't need to convert data type to mysql data type.

## Checklist

- [x] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by StrikeW** [comment]
* **#17813** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/17813?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @StrikeW and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/connector/src/source/cdc/external/postgres.rs [hunk 1]
```diff
@@ -385,29 +362,49 @@ impl PostgresExternalTableReader {
...
+
+                    let order_key = Self::get_order_key(&primary_keys);
+                    let scan_sql = format!(
+                        "SELECT {} FROM {} WHERE {} ORDER BY {} LIMIT {scan_limit}",
+                        self.field_names,
+                        Self::get_normalized_table_name(&table_name),
+                        Self::filter_expression(&primary_keys),
+                        order_key,
+                    );
+                    client.prepare(&scan_sql).await?
```
Comment (fuyufjh) [inline src/connector/src/source/cdc/external/postgres.rs:390]: Every time it prepares once and execute once? Then let's just use simple statements such as Line 406~413

Comment (StrikeW) [reply src/connector/src/source/cdc/external/postgres.rs:390]: prepare for each `scan_limit` chunk, we need the data type from upstream to convert `varchar` for `uuid`.

Comment (xiangjinwu) [reply src/connector/src/source/cdc/external/postgres.rs:390]: Can we make it prepare once but only when backfill not finished yet?

**Review by fuyufjh — APPROVED** [review]

File: src/connector/src/source/cdc/external/postgres.rs (hunk 1 above)
Comment (StrikeW) [reply src/connector/src/source/cdc/external/postgres.rs:390]: It is possible,  but needs to refactor the code for example to call a `init` method before doing the snapshot.
We have set the default value of `scan_limit` to 1000, the overhead is low.

**Pull request merged**